bittensor/_dataset/dataset_impl.py

```diff
@@ -1,536 +1,490 @@
-        save_dataset,
-        max_datasets,
-        no_tokenizer, 
-        num_batches,
-        max_directories
-    ):
-        super().__init__()
-        self.block_size = block_size
+            self, 
+            batch_size: int=8, 
+            sequence_length: int=256,
+            block_size: int = 10000,
+            num_workers: int = 1,
+            dataset_name: List[str]=['ArXiv'], 
+            loop:'asyncio.loop'=None, 
+            no_tokenizer: bool = False,
+            data_dir: str =  os.path.expanduser('~/./bittensor/dataset'),
+            save_dataset : bool = False,
+            load_dataset : bool = True,
+            run_generator:bool=True,
+            buffer_size:int=100,
+            num_batches: int = 100,
+            max_datasets: int = 2,
+            max_directories: int=10):
+
+        """
+        Args:
+            batch_size (int):
+                The size of the batch.
+            sequence_length (int):
+                The length of the seqeunce (tokenwise).
+            block_size (int): 
+                The  size of the text data block which is used to create 
+                multiple samples.
+            num_workers (int):
+                Number of workers for pytorch Dataset.
+            dataset_name (List[str]):
+                List of dataset names to include from the pile.
+            loop ('asyncio.loop'):
+                Asyncio loop for class, defaults to default event loop.
+            no_tokenizer (bool):
+                Do not inlcude tokenizer.
+            data_dir (str):
+                Directory for saving assets.
+            save_dataset (bool):
+                Save the dataset text hashes.
+            load_dataset (bool):
+                Load the dataset text hashes.
+            run_generator (bool):
+                Run the background loop for fecthing data blocks.
+            buffer_size (int):
+                Size of the queue for asynchronously putting data blocks.
+            num_batches: (int):
+                Number of generated batches in epoch.
+            max_datasets: (int):
+                Max number of datasets.
+
+        """
+        self.__infinite_dataset_iterator = None
+        self.dataset_size = 0
         self.batch_size = batch_size
+        self.block_size = block_size
         self.num_workers = num_workers
-        self.tokenizer = bittensor.tokenizer( version = bittensor.__version__ )
+        self.sequence_length = sequence_length
         self.dataset_name = dataset_name
-        self.data_dir = data_dir
-        self.save_dataset = save_dataset
-        self.datafile_size_bound = 262158
+        self.set_event_loop(loop=loop)
+        # if datasets is None then refer to all of the availabe datasets 
         self.max_datasets = max_datasets
-        self.__infinite_dataset_iterator = None
+        if len(self.dataset_name) == 0 or self.dataset_name == 'default':
+            self.dataset_name = self.available_datasets
```

Comment:
I would just raise an exception if `len(datasets) == 0`. And because you're explicitly specifying the type in the function call, i wouldn't allow a user to pass in "default" (unless this is important for legacy reasons in the rest of the codebase, in which case we should at least document this).